src/allocator_defrag.c

```diff
@@ -0,0 +1,484 @@
+        /* Mib of fetch number of current slabs in the bin */
+        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curslabs", j);
+        binfo->info_keys.curr_slabs_key_len = sizeof(binfo->info_keys.curr_slabs_key) / sizeof(size_t);
+        assert(!je_mallctlnametomib(buf, binfo->info_keys.curr_slabs_key, &binfo->info_keys.curr_slabs_key_len));
+        /* Mib of fetch nonfull slabs */
+        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.nonfull_slabs", j);
+        binfo->info_keys.nonfull_slabs_key_len = sizeof(binfo->info_keys.nonfull_slabs_key) / sizeof(size_t);
+        assert(!je_mallctlnametomib(buf, binfo->info_keys.nonfull_slabs_key, &binfo->info_keys.nonfull_slabs_key_len));
+
+        /* Mib of fetch num of alloc op */
+        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.nmalloc", j);
+        binfo->info_keys.nmalloc_key_len = sizeof(binfo->info_keys.nmalloc_key) / sizeof(size_t);
+        assert(!je_mallctlnametomib(buf, binfo->info_keys.nmalloc_key, &binfo->info_keys.nmalloc_key_len));
+        /* Mib of fetch num of dealloc op */
+        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.ndalloc", j);
+        binfo->info_keys.ndealloc_key_len = sizeof(binfo->info_keys.ndealloc_key) / sizeof(size_t);
+        assert(!je_mallctlnametomib(buf, binfo->info_keys.ndealloc_key, &binfo->info_keys.ndealloc_key_len));
+        // verify the reverse map of reg_size to bin index
+        if (jemalloc_quantum == 8) {
+            assert(jeSize2BinIndexLgQ3(binfo->reg_size) == j);
+        } else {
+            assert(jeSize2BinIndexLgQ4(binfo->reg_size) == j);
+        }
+    }
+    arena_bin_conf.nbins = nbins;
+    usage_latest.bins_usage = zcalloc(sizeof(jeBusage) * nbins);
+
+    // get the mib of the per memory pointers query command that is used during defrag scan over memory
+    arena_bin_conf.util_batch_query_key_len = sizeof(arena_bin_conf.util_batch_query_key) / sizeof(size_t);
+    if (je_mallctlnametomib("experimental.utilization.batch_query", arena_bin_conf.util_batch_query_key,
+                            &arena_bin_conf.util_batch_query_key_len)) {
+        // jemalloc version does not support utilization query
+        defrag_supported = 0;
+        return -1;
+    }
+    arena_bin_conf.util_query_key_len = sizeof(arena_bin_conf.util_query_key) / sizeof(size_t);
+    assert(!je_mallctlnametomib("experimental.utilization.query", arena_bin_conf.util_query_key,
+                                &arena_bin_conf.util_query_key_len));
+    // defrag is supported mark it to enable defrag queries
+    defrag_supported = 1;
+    return 0;
+}
+
+/* Total size of consumed meomry in unused regs in small bins (AKA external fragmentation). */
+unsigned long allocatorDefragGetFragSmallbins(void) {
+    unsigned long frag = 0;
+    // todo for frag calculation, should we consider sizes above page size?
```

Comment:
Should we consider this before merging? I'm not sure I understand why we should, but think we should at least fill out this TODO otherwise I doubt we'll remember later :)

Reply:
I will remove the todo. I don't think it needs to be part of the initial PR. I will open a separate issue for followup tasks after this one is merged